core/shared/src/main/scala/scalaz/zio/Semaphore.scala

```diff
@@ -0,0 +1,123 @@
+package scalaz.zio
+
+import scala.collection.immutable.{ Queue => IQueue }
+import scala.util.control.NoStackTrace
+
+abstract class Semaphore {
+
+  import Semaphore.NegativeArgument
+
+  //  def available[E]: IO[E, Long]
+
+  //  def count[E]: IO[E, Long]
+
+  def acquireN(n: Long): IO[NegativeArgument.type, Unit]
+
+  def acquire: IO[Void, Unit] = acquireN(1).attempt[Void].toUnit
+
+  //  def tryAcquireN[E](n: Long): IO[E, Boolean]
+
+  //  def tryAcquire[E]: IO[E, Boolean] = tryAcquireN(1)
+
+  def releaseN(n: Long): IO[NegativeArgument.type, Unit]
+
+  def release: IO[Void, Unit] = releaseN(1).attempt[Void].toUnit
+
+  //  def withPermit[E, A](t: IO[E, A]): IO[E, A]
+
+}
+
+object Semaphore {
+
+  private def assertNonNegative(n: Long): IO[NegativeArgument.type, Unit] =
+    if (n < 0) IO.fail(NegativeArgument) else IO.unit
+
+  // making this an object with no stack trace might not be a good idea for debugging
+  object NegativeArgument extends IllegalArgumentException with NoStackTrace
+
+  private type State = Either[IQueue[(Long, Promise[NegativeArgument.type, Unit])], Long]
+
+  private class ConcreteSemaphore(state: Ref[State]) extends Semaphore {
+
+    private def mkGate[E]: IO[E, Promise[E, Unit]] = Promise.make[E, Unit]
+
+    private def awaitGate[E](entry: (Long, Promise[E, Unit])): IO[E, Unit] =
+      IO.unit[E]
+        .bracket0[Unit] { (_, useOutcome) =>
+          useOutcome match {
+            case None => // None outcome means either cancellation or uncaught exception
+              state.modify {
+                case Left(waiting) => Left(waiting.filter(_ != entry))
+                case Right(m)      => Right(m)
+              }.toUnit
+            case _ =>
+              IO.unit
+          }
+        }(_ => entry._2.get)
+
+    private def openGate[E](entry: (Long, Promise[E, Unit])): IO[E, Unit] =
+      entry._2.complete(()).toUnit
+
+    override def acquireN(requested: Long): IO[NegativeArgument.type, Unit] =
+      assertNonNegative(requested) *>
+        mkGate[NegativeArgument.type].flatMap { gate =>
+          state.modify {
+            case Left(waiting) => Left(waiting :+ (requested -> gate))
+            case Right(available) =>
+              if (requested <= available) Right(available - requested)
+              else Left(IQueue((requested - available) -> gate))
+          }.flatMap {
+            case Left(waiting) =>
+              val entry = waiting.lastOption
```

Comment:
We can eliminate this by ensuring the waiting queue is always `NonEmpty`, where `NonEmpty[F[_], A] = (A,  F[A])` (i.e. a tuple of the first element, and then the possibly empty collection containing the remainder.

Reply:
This is fun on filtering out interrupted ones.